### ISSUE TITLE
Set a file association for files with no extension

### ISSUE THREAD

**Issue by rightaway** [Microsoft/vscode#57939]
```
{
	"files.associations": {
		"*": "mydefaulttype"
	}
}
```

This changes the file association for absolutely all files. Should be possible to set a file association only for files that have no extension. So it would match `somefilename` but not `somefilename.txt` and not even `.somefilename`.

**Comment by bpasero** [comment]
@rightaway does it not work when you set:

```json
{
	"files.associations": {
		"somefilename": "mydefaulttype"
	}
}
```

**Comment by rightaway** [comment]
Yes, that works fine on the files named exactly `somefilename`. But can't figure out any way to set a default file type for files with no extension to work.

**Comment by Frankstar** [comment]
still open?

**Comment by HyiouBear** [comment]
Is it not possible to have a special syntax instead of that "*" which matches every extensions?

**Comment by stickycode** [comment]
/[^.]+/ maybe allow regular expressions?

Or how about reading the header, shell scripts often don't have an extension but they do have a header e.g. #!/bin/sh or #!/usr/bin/env bash

**Comment by sam-rad** [comment]
According to [VSCode Programming Languages Documentation](https://code.visualstudio.com/docs/languages/overview#_can-i-map-additional-file-extensions-to-a-language) the glob pattern should be supported, right?

But I wonder why these don't work:
```json
"files.associations": {
    "**/!(*.*)": "markdown",
    "**/+([^.])": "markdown"
  }
```

You can test them on [Digital Ocean | Glob Tool](https://www.digitalocean.com/community/tools/glob?glob=%2A%2A%2F%21%28%2A.%2A%29&tests=a-file-with-no-extension&tests=a%2Fb%2Fc%2Fd%2Fe%2Fa-file-with-no-extension&tests=blah.txt&tests=a%2Fb%2Fc%2Fd%2Fe%2Fblah.txt)

**Comment by mjy9088** [comment]
Did you try `files.defaultLanguage`?

This changes other files' too, but I think it might help you

**Comment by mjy9088** [comment]
It's seems to [not possible with glob only](https://stackoverflow.com/a/52573274).

Do you have suggestions on how to represent something without an extension, in `settings.json`?

Example - Add PCRE regex support, and specify the pattern type

```javascript
"files.associations": {
    "**/page/*.js": "javascriptreact", // default: glob
    ".*/docs/[^\\.]*$": { "pattern": "regex", "language": "markdown" }, // pattern type specified?
}
```

IMO this way is straightforward but looks not so good

**Comment by ThomasReulen** [comment]
Since I also ran into this issue, I will provide my configuration for this. 

I wanted to have plaintext for all files without extension. So i just define all files as plaintext as default, and then overwrite specific languages with the correct setting. You can do this easily by opening a new file with an extension not yet in the list. It will be marked as plain text. Then click in the bottom status bar on "Plain Text" and "Configure File Association for .<ext>..." and select the language from the list. The user settings will be updated automatically and from this point onwards the file extension will be recognized correctly.

```

    "files.associations": {
        "**/*": "plaintext",
        "*.go": "go",
        "*.{htm,html}": "html",
        "*.{js,ts}": "javascript",
        "*.py": "python",
        "*.sh": "sh",
        "*.json": "json",
        "*.md": "markdown",
        "*.ps1": "powershell"
    },
```

Hope this is helpful for someone :-) Cheers

**Comment by goyalyashpal** [comment]
> ```json
> "somefilename": "mydefaulttype"
> ```

hey! how can i know the code for "mydefaulttype" - it's obvious for lots of stuff, but many times it's not.

For example: "Git Commit Message" was shown in the status bar, but the code for it was "git-commit" which i found out after jumping various hoops.

what would have been ur way to find out that code?

**Comment by DarrenCook** [comment]
I've had to resort to the suggestion of ThomasReulen above. A shame we have to dumb down vscode and will have to manually add a new entry each time we come to a language extension for the first time. (E.g. I just had to add `"*.jsonl":"jsonl"`)

Regarding https://github.com/microsoft/vscode/issues/57939#issuecomment-1033993749 one technique we've used to mix regex and plain filters like this is if it starts and ends in `/` then treat it as a regex; though not ideal here because it still looks glob-ish.

But can I suggest just adding no-extension as a special case? Ideas:
1. Allow `Item` to be blank, and have that mean no extension
2. Use a special character. E.g. if item is a single `_` it means no extension.
3. Add `files.noExtAssociation` as a new setting.

**Comment by Jbleezy** [comment]
Why does `!(*.*)` not work here?
I use the VSCode extension `JatinSanghvi.color-my-text` and that glob pattern works there to get only files with no extension.

**Comment by mjy9088** [comment]
I don't know why, but `!(**/*.*)`, `!(*.*)`, and many variants I've tried just not work

**Comment by TriMoon** [comment]
I just tried `"shell_aliases.d/[^.]+": "shellscript"` and thats not working either 🤷‍♀
